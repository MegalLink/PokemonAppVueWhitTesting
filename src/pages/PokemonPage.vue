<template>
  <h1 v-if="!pokemon">Espere porfavor</h1>
  <div v-else>
    <h1>¿Quien es este pokémon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :reveal="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selectedPokemon="checkAnswer" />
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">Nuevo juego</button>
    </template>
  </div>
</template>
<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
    };
  },
  mounted() {
    this.getPokemonArray();
  },
  methods: {
    async getPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },
    checkAnswer(selectedId) {
      this.showPokemon = true;
      this.showAnswer = true;
      if (selectedId == this.pokemon.id) {
        this.message = `Correcto, es ${this.pokemon.name}`;
      } else {
        this.message = `Nop, es ${this.pokemon.name}`;
      }
    },
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemon = null;
      this.getPokemonArray();
    },
  },
};
</script>
