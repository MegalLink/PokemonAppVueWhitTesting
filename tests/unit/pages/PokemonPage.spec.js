import { shallowMount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import pokemonsMock from "../mocks/pokemons.mock";
describe("Test PokemonPage", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });
  it("should match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should call getPokemonArray on mount", () => {
    //Como es de un mount el spy debemos ponerlo antes por que todavia no ha sido llamado,
    const mixPokemonArraySpy = jest.spyOn(
      PokemonPage.methods,
      "getPokemonArray"
    );
    //Ejecutamos el componente
    shallowMount(PokemonPage);

    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  it("should call getPokemonArray on mount", () => {
    //Ejecutamos el componente
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemonsMock,
          pokemon: pokemonsMock[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });
  });

  it("should existe PokemonPicture and PokemonOptions Components", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemonsMock,
          pokemon: pokemonsMock[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    const optionsComponent = wrapper.find("pokemon-options-stub");
    const pictureComponent = wrapper.find("pokemon-picture-stub");
    expect(pictureComponent.exists()).toBeTruthy();
    expect(optionsComponent.exists()).toBeTruthy();
    expect(pictureComponent.attributes("pokemonid")).toBe("100");
    expect(optionsComponent.attributes("pokemons")).toBeTruthy();
  });

  it("should change showPokemon,showAnser to true and change message when checkAnswer function is triggered", async () => {
    const pokemonMatch = pokemonsMock[1];
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemonsMock,
          pokemon: pokemonMatch,
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    //wrapper.vm i so powerfull :D
    await wrapper.vm.checkAnswer(pokemonMatch.id);
    const h2Message = wrapper.find("h2");
    expect(h2Message.exists()).toBeTruthy();
    expect(h2Message.text()).toBe(`Correcto, es ${pokemonMatch.name}`);
    expect(wrapper.vm.showPokemon).toBeTruthy();
    expect(wrapper.vm.showAnswer).toBeTruthy();
    await wrapper.vm.checkAnswer(123);
    expect(h2Message.text()).toBe(`Nop, es ${pokemonMatch.name}`);
  });
});
