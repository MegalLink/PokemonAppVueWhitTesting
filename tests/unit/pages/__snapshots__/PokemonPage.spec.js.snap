// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PokemonPage should match snapshot 1`] = `<h1>Espere porfavor</h1>`;
